File: apps/scraper/app/json_ld.py
import json
import logging
import re

from bs4 import BeautifulSoup

from .parsing import ExtractionResult, _parse_price_match

logger = logging.getLogger(__name__)


def _extract_json_ld(soup: BeautifulSoup) -> ExtractionResult | None:
    """Extract price and availability from JSON-LD Product schema."""
    for script in soup.find_all("script", type="application/ld+json"):
        try:
            data = json.loads(script.string or "")
        except (json.JSONDecodeError, TypeError):
            continue

        products = _find_json_ld_products(data)
        for product in products:
            result = _extract_from_product_json_ld(product)
            if result is not None:
                return result

    return None


def _find_json_ld_products(data) -> list[dict]:
    """Recursively find all Product objects in JSON-LD data."""
    products = []
    if isinstance(data, list):
        for item in data:
            products.extend(_find_json_ld_products(item))
        return products

    if not isinstance(data, dict):
        return products

    type_val = data.get("@type", "")
    types = type_val if isinstance(type_val, list) else [type_val]
    if "Product" in types:
        products.append(data)

    graph = data.get("@graph")
    if isinstance(graph, list):
        for item in graph:
            products.extend(_find_json_ld_products(item))

    for key in ("mainEntity", "mainEntityOfPage"):
        nested = data.get(key)
        if isinstance(nested, dict):
            products.extend(_find_json_ld_products(nested))

    return products


def _extract_from_product_json_ld(product: dict) -> ExtractionResult | None:
    """Extract price/stock from a single JSON-LD Product object."""
    offers_raw = product.get("offers", {})

    if isinstance(offers_raw, dict):
        offer_type = offers_raw.get("@type", "")
        offer_types = offer_type if isinstance(offer_type, list) else [offer_type]
        if "AggregateOffer" in offer_types:
            return _extract_aggregate_offer(offers_raw)
        offers = [offers_raw]
    elif isinstance(offers_raw, list):
        offers = offers_raw
    else:
        return None

    return _pick_best_offer(offers)


def _extract_aggregate_offer(agg: dict) -> ExtractionResult | None:
    """Extract from AggregateOffer — prefer lowPrice, fall back to nested offers."""
    raw_parts = []
    price = None
    stock_status = None

    for key in ("lowPrice", "price"):
        val = agg.get(key)
        if val is not None:
            price = _parse_json_ld_price(val)
            if price is not None:
                raw_parts.append(f"{key}={price}")
                break

    stock_status = _parse_json_ld_availability(agg.get("availability", ""))
    if stock_status:
        raw_parts.append(f"availability={stock_status}")

    if price is not None or stock_status is not None:
        return ExtractionResult(price=price, stock_status=stock_status, raw=", ".join(raw_parts))

    nested = agg.get("offers")
    if isinstance(nested, list):
        return _pick_best_offer(nested)

    return None


def _pick_best_offer(offers: list[dict]) -> ExtractionResult | None:
    """Pick the best offer: prefer lowest in-stock, fall back to lowest overall."""
    in_stock_prices = []
    all_prices = []

    for offer in offers:
        price = _parse_json_ld_price(offer.get("price"))
        if price is None:
            price = _parse_json_ld_price(offer.get("lowPrice"))
        availability = _parse_json_ld_availability(offer.get("availability", ""))

        if price is not None:
            all_prices.append((price, availability, offer))
            if availability == "in_stock":
                in_stock_prices.append((price, availability, offer))

    chosen = None
    if in_stock_prices:
        chosen = min(in_stock_prices, key=lambda x: x[0])
    elif all_prices:
        chosen = min(all_prices, key=lambda x: x[0])

    if chosen is None:
        if offers:
            avail = _parse_json_ld_availability(offers[0].get("availability", ""))
            if avail:
                return ExtractionResult(price=None, stock_status=avail, raw=f"availability={avail}")
        return None

    price, stock_status, _ = chosen
    raw_parts = [f"price={price}"]
    if stock_status:
        raw_parts.append(f"availability={stock_status}")
    return ExtractionResult(price=price, stock_status=stock_status, raw=", ".join(raw_parts))


def _parse_json_ld_price(val) -> float | None:
    """Parse a JSON-LD price value, stripping currency symbols."""
    if val is None:
        return None
    raw = re.sub(r"[^\d.,]", "", str(val))
    return _parse_price_match(raw)


def _parse_json_ld_availability(val) -> str | None:
    """Map JSON-LD availability to 'in_stock' or 'out_of_stock'."""
    s = str(val).lower()
    if "instock" in s:
        return "in_stock"
    if "outofstock" in s:
        return "out_of_stock"
    if "preorder" in s or "presale" in s:
        return "in_stock"
    if "discontinued" in s or "soldout" in s:
        return "out_of_stock"
    return None

File: apps/scraper-rs/src/json_ld.rs
use scraper::Html;
use serde_json::Value;

use crate::{
    dom::select_all,
    parsing::{ExtractionResult, parse_price_match, text_of_element},
};

pub fn extract_json_ld(document: &Html) -> Option<ExtractionResult> {
    for script in select_all(document, r#"script[type="application/ld+json"]"#) {
        let raw = text_of_element(&script);
        let Ok(value) = serde_json::from_str::<Value>(&raw) else {
            continue;
        };

        let mut products = Vec::new();
        find_products(&value, &mut products);
        for product in products {
            if let Some(result) = extract_from_product_json_ld(product) {
                return Some(result);
            }
        }
    }

    None
}

fn find_products<'a>(value: &'a Value, products: &mut Vec<&'a Value>) {
    match value {
        Value::Array(items) => {
            for item in items {
                find_products(item, products);
            }
        }
        Value::Object(map) => {
            let types = match map.get("@type") {
                Some(Value::String(value)) => vec![value.as_str()],
                Some(Value::Array(values)) => values.iter().filter_map(Value::as_str).collect(),
                _ => Vec::new(),
            };
            if types.iter().any(|value| *value == "Product") {
                products.push(value);
            }

            if let Some(Value::Array(graph)) = map.get("@graph") {
                for item in graph {
                    find_products(item, products);
                }
            }

            for key in ["mainEntity", "mainEntityOfPage"] {
                if let Some(nested) = map.get(key) {
                    find_products(nested, products);
                }
            }
        }
        _ => {}
    }
}

fn extract_from_product_json_ld(product: &Value) -> Option<ExtractionResult> {
    let offers = product.get("offers")?;

    match offers {
        Value::Object(map) => {
            let offer_types = match map.get("@type") {
                Some(Value::String(value)) => vec![value.as_str()],
                Some(Value::Array(values)) => values.iter().filter_map(Value::as_str).collect(),
                _ => Vec::new(),
            };

            if offer_types.iter().any(|value| *value == "AggregateOffer") {
                return extract_aggregate_offer(offers);
            }

            pick_best_offer(&[offers])
        }
        Value::Array(values) => {
            let refs = values.iter().collect::<Vec<_>>();
            pick_best_offer(&refs)
        }
        _ => None,
    }
}

fn extract_aggregate_offer(aggregate_offer: &Value) -> Option<ExtractionResult> {
    let mut price = None;
    let mut stock_status = None;
    let mut raw_parts = Vec::new();

    for key in ["lowPrice", "price"] {
        if let Some(value) = aggregate_offer.get(key).and_then(parse_json_ld_price) {
            price = Some(value);
            raw_parts.push(format!("{key}={value}"));
            break;
        }
    }

    if let Some(availability) = aggregate_offer
        .get("availability")
        .and_then(parse_json_ld_availability)
    {
        raw_parts.push(format!("availability={availability}"));
        stock_status = Some(availability);
    }

    if price.is_some() || stock_status.is_some() {
        return Some(ExtractionResult {
            price,
            stock_status,
            raw: Some(raw_parts.join(", ")),
        });
    }

    aggregate_offer
        .get("offers")
        .and_then(Value::as_array)
        .map(|values| values.iter().collect::<Vec<_>>())
        .and_then(|offers| pick_best_offer(&offers))
}

fn pick_best_offer(offers: &[&Value]) -> Option<ExtractionResult> {
    let mut in_stock_prices = Vec::new();
    let mut all_prices = Vec::new();

    for offer in offers {
        let price = offer
            .get("price")
            .and_then(parse_json_ld_price)
            .or_else(|| offer.get("lowPrice").and_then(parse_json_ld_price));
        let availability = offer
            .get("availability")
            .and_then(parse_json_ld_availability);

        if let Some(price) = price {
            if availability.as_deref() == Some("in_stock") {
                in_stock_prices.push((price, availability.clone()));
            }
            all_prices.push((price, availability));
        }
    }

    let chosen = in_stock_prices
        .into_iter()
        .min_by(|left, right| left.0.total_cmp(&right.0))
        .or_else(|| {
            all_prices
                .into_iter()
                .min_by(|left, right| left.0.total_cmp(&right.0))
        });

    if let Some((price, stock_status)) = chosen {
        let mut raw_parts = vec![format!("price={price}")];
        if let Some(availability) = stock_status.clone() {
            raw_parts.push(format!("availability={availability}"));
        }
        return Some(ExtractionResult {
            price: Some(price),
            stock_status,
            raw: Some(raw_parts.join(", ")),
        });
    }

    offers.first().and_then(|offer| {
        offer
            .get("availability")
            .and_then(parse_json_ld_availability)
            .map(|availability| ExtractionResult {
                price: None,
                stock_status: Some(availability.clone()),
                raw: Some(format!("availability={availability}")),
            })
    })
}

fn parse_json_ld_price(value: &Value) -> Option<f64> {
    let raw = match value {
        Value::String(value) => value.as_str(),
        Value::Number(value) => {
            return value
                .as_f64()
                .and_then(|price| parse_price_match(&price.to_string()));
        }
        _ => return None,
    };

    parse_price_match(raw)
}

fn parse_json_ld_availability(value: &Value) -> Option<String> {
    let normalized = value.to_string().to_lowercase();
    if normalized.contains("instock")
        || normalized.contains("preorder")
        || normalized.contains("presale")
    {
        return Some("in_stock".to_string());
    }
    if normalized.contains("outofstock")
        || normalized.contains("discontinued")
        || normalized.contains("soldout")
    {
        return Some("out_of_stock".to_string());
    }
    None
}
